### PR TITLE
Fix scheduler ws race condition

### PR DESCRIPTION
**Pull request by Martchus**
This change is again a little bit bigger for "just" fixing a race condition. However, it involves changing the protocol a little bit (see commit messages for a description of the actual changes).

But now there should be tests for all relevant communication bits between the worker and the web UI so this likely works as expected. However, the race condition I'm trying to fix here might have become even worse by previously [fixing another race condition](https://github.com/os-autoinst/openQA/commit/f3ea708082fac95684ca5f925bb356ced1c1705e). So who knows which race condition I'm now overlooking. But when this is merged at least 2 problems should be finally addressed:

1. The worker takes the job faster than the web UI fully processes the job-worker assignment. (That was https://github.com/os-autoinst/openQA/commit/f3ea708082fac95684ca5f925bb356ced1c1705e.)
2. The worker takes the job tool slow and the web UI already "re-schedules" the job. (This PR, see https://progress.opensuse.org/issues/62984#note-19).

### CODE REVIEW

File: lib/OpenQA/Worker/CommandHandler.pm [hunk 1]
```diff
@@ -67,18 +66,12 @@ sub handle_command {
             return undef;
         }
     }
-    elsif ($type ne 'info') {
+    elsif ($type ne 'info' && $type ne 'grab_job' && $type ne 'grab_jobs') {
```
Comment (kalikiana) [inline lib/OpenQA/Worker/CommandHandler.pm]: Maybe the condition should be the types we want?

File: t/05-scheduler-full.t [hunk 2]
```diff
@@ -129,9 +130,17 @@ subtest 're-scheduling and incompletion of jobs when worker is unresponsive or c
...
-    my $unstable_w_pid = unresponsive_worker($k->key, $k->secret, "http://localhost:$mojoport", 3);
+    # simulate a broken worker; it will register itself but declare itself as broken
+    my $broken_w_pid = broken_worker($k->key, $k->secret, "http://localhost:$mojoport", 3, 'out of order');
+    wait_for_worker($schema, 5);
+    $allocated = scheduler_step();
+    is(@$allocated, 0, 'scheduler does not consider broken worker for allocating job');
+    kill_service($broken_w_pid, 1);
+
+    # simulate a rejective worker; it looks free but rejects all jobs assigned to it
+    my $rejective_w_pid = rejective_worker($k->key, $k->secret, "http://localhost:$mojoport", 3, 'rejection reason');
```
Comment (kalikiana) [inline t/05-scheduler-full.t]: Before:
 - unresponsive worker, doesn't grab any jobs

After:
- broken worker, declares itself as broken
- rejective worker, rejects all jobs

I'm not clear on the two new ones. What is "broken"?
Also... "rejective" is not a word... did you mean repudiative?

File: t/05-scheduler-full.t [hunk 3]
```diff
@@ -154,10 +163,10 @@ subtest 're-scheduling and incompletion of jobs when worker is unresponsive or c
         sleep .2;
     }
     ok($job_scheduled, 'assigned job set back to scheduled if worker reports back again but has abandoned the job');
-    kill_service($unstable_w_pid, 1);
+    kill_service($rejective_w_pid, 1);
 
-    # start unstable worker again
-    $unstable_w_pid = unstable_worker($k->key, $k->secret, "http://localhost:$mojoport", 3, -1);
+    # start an unatable worker which will basically just ignore the assignment
```
Comment (kalikiana) [inline t/05-scheduler-full.t]: `# start an unstable worker...`?
```suggestion
    # start an unstable worker which will basically just ignore the assignment
```

File: t/05-scheduler-full.t (hunk 2 above)
Comment (Martchus) [reply t/05-scheduler-full.t]: Broken means the worker is in the "broken" state. That's actually a well-defined work within the scope of the openQA worker. It is also already used within the UI for the that particular worker status.

Google translator says "rejective" means "ablehnend" and that's exactly what I mean here. And lexico.cpm has a matching definition for it, too. So that worker rejects doing any work (sends the *reject* message introduced by this PR). That is different from the unresponsive worker which does not respond at all (just ignores messages but does not send the *reject* message).

File: lib/OpenQA/Worker/CommandHandler.pm (hunk 1 above)
Comment (Martchus) [reply lib/OpenQA/Worker/CommandHandler.pm]: I had that thought as well but didn't want to re-decide about that design for the sake of this change. The current design makes basically the assumption that all message types are job-specific by default and other messages are the exception. I had to make grab job messages the exception as well because these need to be handled separately (they are job-specific but refer to a *new* job rather than the current one).

File: t/05-scheduler-full.t (hunk 2 above)
Comment (kalikiana) [reply t/05-scheduler-full.t]: Interesting. `lexico.com` shows me the verb or the adjective depending on how fast I type.

I think what's confusing me here is the mix. Say `# Simulate a worker in broken state` would clearly refer to the status string you get. But in the second case it should describe the behavior, like `# Simulate an idle worker that rejects all jobs assigned to it`.
How would that sound to you?

File: t/lib/OpenQA/Test/Utils.pm [hunk 4]
```diff
@@ -395,16 +407,28 @@ sub c_worker {
                 handle_command => sub {
                     my ($self, $tx, $json) = @_;
                     log_debug('Received ws message: ' . Dumper($json));
+
+                    # if we've got a single job ID and a rejection reason simulate a worker
+                    # which rejects the job
+                    my $rejection_reason = $options{rejection_reason};
+                    return undef unless defined $rejection_reason;
+                    my $job_id = (($json // {})->{job} // {})->{id};
```
Comment (kraih) [inline t/lib/OpenQA/Test/Utils.pm]: Would autovivification be harmuful here?

File: lib/OpenQA/Schema/Result/Workers.pm [hunk 5]
```diff
@@ -279,4 +280,34 @@ sub set_current_job {
...
+
+        # consider only assigned jobs here
+        # note: Running jobs are only marked as incomplete on worker registration (and not here) because that
+        #       operation can be quite costly.
+        next if $associated_job->state ne ASSIGNED;
+
+        # set associated job which was only assigned back to scheduled
+        # note: Using a transaction here so we don't end up with an inconsistent state when an error occurs.
+        try {
+            $self->result_source->schema->txn_do(sub { $associated_job->reschedule_state; });
```
[the remark this answers is not in the document]
Comment (Martchus) [reply lib/OpenQA/Schema/Result/Workers.pm]: I'm often writing C++ as well where the `;` is mandatory so I'm more used to it. But I have no preference so I'll remove it as it apparently not Perl-ish.

File: t/lib/OpenQA/Test/Utils.pm (hunk 4 above)
Comment (Martchus) [reply t/lib/OpenQA/Test/Utils.pm]: I guess I've forgotten about that super cool Perl feature when writing this code. I suppose it would actually be quite useful here :-)

File: lib/OpenQA/Worker/CommandHandler.pm (hunk 1 above)
Comment (Martchus) [reply lib/OpenQA/Worker/CommandHandler.pm]: Changed it to `exists $COMMANDS_SPECIFIC_TO_CURRENT_JOB{$type}` so it is also clear what these types have in common.

File: t/24-worker-webui-connection.t [hunk 6]
```diff
@@ -455,15 +493,61 @@ qr/Ignoring WS message from http:\/\/test-host with type livelog_stop and job ID
...
+    );
+    combined_like(
+        sub {
+            $command_handler->handle_command(
+                undef,
+                {
+                    type     => 'grab_jobs',
+                    job_info => {sequence => 'not an array', data => {42 => 'foo'}},
+                });
+        },
```
Comment (okurz) [inline t/24-worker-webui-connection.t]: just regarding syntax, this looks pretty verbose. Maybe that can be extracted into a function?

File: t/24-worker-webui-connection.t [hunk 7]
```diff
@@ -455,15 +493,61 @@ qr/Ignoring WS message from http:\/\/test-host with type livelog_stop and job ID
...
         sub { $command_handler->handle_command(undef, {type => 'foo'}); },
         qr/Ignoring WS message with unknown type foo.*/,
         'ignoring messages of unknown type',
     );
 
+    is_deeply(
+        $ws->sent_messages,
+        [
+            {json => {job_ids => [42], reason => 'some error', type => 'rejected'}},
+            {json => {job_ids => [42], reason => 'some error', type => 'rejected'}},
```
Comment (okurz) [inline t/24-worker-webui-connection.t]: probably better to extract the hash into a variable to be reused

File: t/24-worker-webui-connection.t (hunk 6 above)
Comment (kraih) [reply t/24-worker-webui-connection.t]: It's verbose, but also well structured.

**Comment by Martchus** [comment]
I added unit tests for the corner cases (which are otherwise hard to reproduce). Of course I've tested the usual test execution locally, too. Testing this on a staging setup makes not much sense (it makes it even harder to test the corner cases). The most useful additional testing which comes to my mind would be a manual fullstack test with a worker which is connected to two web UIs (the code for this is tested in unit/integration tests). But I actually think this is ready to be merged.

**Comment by AdamWill** [comment]
I'll roll this out to our staging setup tomorrow and keep an eye on it. I should be able to report back if it seems to be reducing/eliminating the occurrence of incompletions.

**Comment by AdamWill** [comment]
ok, now I've seen #2772, maybe I won't :)

**Comment by okurz** [comment]
yes. I have reverted with #2772 so that master can be considered stable again.
